flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,411 @@
+		}
+
+		channel.runPendingTasks();
+
+		List<NettyMessage> decodedMessages = new ArrayList<>();
+		Object input;
+		while ((input = channel.readInbound()) != null) {
+			assertTrue(input instanceof NettyMessage);
+			decodedMessages.add((NettyMessage) input);
+		}
+
+		return decodedMessages;
+	}
+
+	private void verifyDecodedMessages(NettyChannelAndInputChannelIds context, NettyMessage[] expected, List<NettyMessage> decodedMessages) {
+		assertEquals(expected.length, decodedMessages.size());
+		for (int i = 0; i < expected.length; ++i) {
+			assertEquals(expected[i].getClass(), decodedMessages.get(i).getClass());
+
+			if (expected[i] instanceof NettyMessage.BufferResponse) {
+				BufferResponse expectedBufferResponse = (BufferResponse) expected[i];
+				BufferResponse decodedBufferResponse = (BufferResponse) decodedMessages.get(i);
+
+				assertEquals(expectedBufferResponse.backlog, decodedBufferResponse.backlog);
+				assertEquals(expectedBufferResponse.sequenceNumber, decodedBufferResponse.sequenceNumber);
+				assertEquals(expectedBufferResponse.bufferSize, decodedBufferResponse.bufferSize);
+				assertEquals(expectedBufferResponse.receiverId, decodedBufferResponse.receiverId);
+
+				if (expectedBufferResponse.bufferSize == 0 ||
+					expectedBufferResponse.receiverId.equals(context.getReleasedChannelId()) ||
+					expectedBufferResponse.receiverId.equals(context.getRemovedChannelId())) {
+
+					assertNull(decodedBufferResponse.getBuffer());
+				} else if (expectedBufferResponse.receiverId.equals(context.getNormalChannelId())) {
+					assertEquals(expectedBufferResponse.getBuffer(), decodedBufferResponse.getBuffer());
+				} else {
+					fail("The decoded buffer response was sent to unknown input channel.");
+				}
+			} else if (expected[i] instanceof NettyMessage.ErrorResponse) {
+				ErrorResponse expectedErrorResponse = (ErrorResponse) expected[i];
+				ErrorResponse decodedErrorResponse = (ErrorResponse) decodedMessages.get(i);
+
+				assertEquals(expectedErrorResponse.receiverId, decodedErrorResponse.receiverId);
+				assertEquals(expectedErrorResponse.cause.getClass(), decodedErrorResponse.cause.getClass());
+				assertEquals(expectedErrorResponse.cause.getMessage(), decodedErrorResponse.cause.getMessage());
+			} else {
+				fail("Unsupported message type");
+			}
+		}
+	}
+
+	private void releaseBuffers(ByteBuf[] buffers) {
+		for (ByteBuf buffer : buffers) {
+			buffer.release();
+		}
+	}
+
+	private Buffer createDataBuffer(int size) {
+		MemorySegment segment = MemorySegmentFactory.allocateUnpooledSegment(size);
+		NetworkBuffer buffer = new NetworkBuffer(segment, FreeingBufferRecycler.INSTANCE);
+		for (int i = 0; i < size / 4; ++i) {
+			buffer.writeInt(i);
+		}
+
+		return buffer;
+	}
+
+	private NettyChannelAndInputChannelIds createPartitionRequestClientHandler(int numberOfBuffersInNormalChannel) throws IOException, InterruptedException {
+		NetworkBufferPool networkBufferPool = new NetworkBufferPool(
```

Comment:
maybe we can make it in class level in `setup` if really necessary, then we can confirm to release it in `teardown`

Reply:
Change input channel into specialized implementation, thus the problem does not exist.